frontend/model/contracts/chatroom.js

```diff
@@ -327,36 +334,94 @@ sbp('chelonia/defineContract', {
         emitMessageEvent({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
       }
     },
-    'gi.contracts/chatroom/deleteMessage': {
-      validate: objectMaybeOf({
-
-      }),
-      process ({ data, meta }, { state }) {
-
-      }
-    },
     'gi.contracts/chatroom/editMessage': {
-      validate: objectMaybeOf({
-
-      }),
+      validate: (data, { state, meta }) => {
+        objectOf({
+          id: string,
+          text: string
+        })(data)
+        // TODO: need to check if the meta.username === message.from
+      },
       process ({ data, meta }, { state }) {
-
+        for (let i = state.messages.length - 1; i >= 0; i--) {
+          if (state.messages[i].id === data.id) {
```

Comment:
I agree with @snowteamer  - this pattern is showing up in your code in several places, so let's DRY this.

I also agree with you @Silver-IT, that it's better to start the search from the end instead of the beginning, and so we should implement our own version of `.find`, e.g. something like:

```js
function findMessageIdx(id, messages) {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i].id === id) {
      return i
    }
  }
  return -1
}
```

Then you can remove the loops you have everywhere else.

Reply:
Nice catch. 10Q